src/include/cephfs/metrics/Types.h

```diff
@@ -0,0 +1,284 @@
+  }
+
+  void decode(bufferlist::const_iterator &iter) {
+    using ceph::decode;
+    DECODE_START(1, iter);
+    decode(lat, iter);
+    DECODE_FINISH(iter);
+  }
+
+  void dump(Formatter *f) const {
+    f->dump_int("latency", lat);
+  }
+};
+
+struct WriteLatencyPayload {
+  static const ClientMetricType METRIC_TYPE = ClientMetricType::CLIENT_METRIC_TYPE_WRITE_LATENCY;
+
+  utime_t lat;
+
+  WriteLatencyPayload() { }
+  WriteLatencyPayload(utime_t lat)
+    : lat(lat) {
+  }
+
+  void encode(bufferlist &bl) const {
+    using ceph::encode;
+    ENCODE_START(1, 1, bl);
+    encode(lat, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator &iter) {
+    using ceph::decode;
+    DECODE_START(1, iter);
+    decode(lat, iter);
+    DECODE_FINISH(iter);
+  }
+
+  void dump(Formatter *f) const {
+    f->dump_int("latency", lat);
+  }
+};
+
+struct MetadataLatencyPayload {
+  static const ClientMetricType METRIC_TYPE = ClientMetricType::CLIENT_METRIC_TYPE_METADATA_LATENCY;
+
+  utime_t lat;
+
+  MetadataLatencyPayload() { }
+  MetadataLatencyPayload(utime_t lat)
+    : lat(lat) {
+  }
+
+  void encode(bufferlist &bl) const {
+    using ceph::encode;
+    ENCODE_START(1, 1, bl);
+    encode(lat, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator &iter) {
+    using ceph::decode;
+    DECODE_START(1, iter);
+    decode(lat, iter);
+    DECODE_FINISH(iter);
+  }
+
+  void dump(Formatter *f) const {
+    f->dump_int("latency", lat);
+  }
+};
+
+struct UnknownPayload {
+  static const ClientMetricType METRIC_TYPE = static_cast<ClientMetricType>(-1);
+
+  UnknownPayload() { }
+
+  void encode(bufferlist &bl) const {
+  }
+
+  void decode(bufferlist::const_iterator &iter) {
+  }
+
+  void dump(Formatter *f) const {
+  }
+};
+
+typedef boost::variant<CapInfoPayload,
+                       ReadLatencyPayload,
+                       WriteLatencyPayload,
+                       MetadataLatencyPayload,
+                       UnknownPayload> ClientMetricPayload;
+
+// metric update message sent by clients
+struct ClientMetricMessage {
+public:
+  ClientMetricMessage(const ClientMetricPayload &payload = UnknownPayload())
+    : payload(payload) {
+  }
+
+  class EncodePayloadVisitor : public boost::static_visitor<void> {
+  public:
+    explicit EncodePayloadVisitor(bufferlist &bl) : m_bl(bl) {
+    }
+
+    template <typename ClientMetricPayload>
+    inline void operator()(const ClientMetricPayload &payload) const {
+      using ceph::encode;
+      encode(static_cast<uint32_t>(ClientMetricPayload::METRIC_TYPE), m_bl);
+      payload.encode(m_bl);
+    }
+
+  private:
+    bufferlist &m_bl;
+  };
+
+  class DecodePayloadVisitor : public boost::static_visitor<void> {
+  public:
+    DecodePayloadVisitor(bufferlist::const_iterator &iter) : m_iter(iter) {
+    }
+
+    template <typename ClientMetricPayload>
+    inline void operator()(ClientMetricPayload &payload) const {
+      using ceph::decode;
```

Comment:
maybe this could be removed.

Reply:
Thanks for bringing this up -- this is done is every other encode/decode routine in src/include (and elsewhere too). We should clean all this up in a separate PR.

Reply:
Yeah, make sense.